src/main/java/edu/cornell/kfs/coa/document/validation/impl/SubAccountGlobalRule.java

```diff
@@ -444,13 +457,371 @@ protected boolean checkCgIcrRules(String subAccountTypeCode, SubAccountGlobalDet
+    }
+    
+    protected boolean accountIsForContractsAndGrants(SubAccountGlobalNewAccountDetail newAccountDetail) {
+        newAccountDetail.refreshReferenceObject(KFSPropertyConstants.ACCOUNT);
+        Account account = newAccountDetail.getAccount();
+        if (ObjectUtils.isNotNull(account)) {
+            account.refreshReferenceObject(KFSPropertyConstants.SUB_FUND_GROUP);
+            SubFundGroup subFundGroup = account.getSubFundGroup();
+            if (ObjectUtils.isNotNull(subFundGroup)) {
+                return getSubFundGroupService().isForContractsAndGrants(subFundGroup);
+            }
+        }
+        return false;
+    }
+    
+    protected boolean checkIndirectCostRecoveryFieldsHaveValuesForCgExpenseAccounts(SubAccountGlobal subAccountGlobal) {
+        boolean success = true;
+        if (checkCgIcrIsEmpty()) {
+            putGlobalError(CUKFSKeyConstants.ERROR_DOCUMENT_SUB_ACCOUNT_GLOBAL_CG_EXPENSE_REQUIRED_SECTION,
+                    CUKFSConstants.SUB_ACCOUNT_GLOBAL_CG_ICR_SECTION_NAME);
+            success = false;
+        }
+        if (subAccountGlobal.getIndirectCostRecoveryAccounts().isEmpty()) {
+            putGlobalError(CUKFSKeyConstants.ERROR_DOCUMENT_SUB_ACCOUNT_GLOBAL_CG_EXPENSE_REQUIRED_SECTION,
+                    CUKFSConstants.SUB_ACCOUNT_GLOBAL_CG_ICR_ACCOUNTS_SECTION_NAME);
+            success = false;
+        }
+        return success;
+    }
+    
+    protected boolean checkCgFieldsAreEmptyForNonCgAccounts(SubAccountGlobal subAccountGlobal) {
+        boolean success = true;
+        if (!checkCgCostSharingIsEmpty()) {
+            putGlobalError(CUKFSKeyConstants.ERROR_DOCUMENT_SUB_ACCOUNT_GLOBAL_NON_CG_NON_BLANK_SECTION,
+                    CUKFSConstants.SUB_ACCOUNT_GLOBAL_CG_COST_SHARE_SECTION_NAME);
+            success = false;
+        }
+        if (!checkCgIcrIsEmpty()) {
+            putGlobalError(CUKFSKeyConstants.ERROR_DOCUMENT_SUB_ACCOUNT_GLOBAL_NON_CG_NON_BLANK_SECTION,
+                    CUKFSConstants.SUB_ACCOUNT_GLOBAL_CG_ICR_SECTION_NAME);
+            success = false;
+        }
+        if (!subAccountGlobal.getIndirectCostRecoveryAccounts().isEmpty()) {
+            putGlobalError(CUKFSKeyConstants.ERROR_DOCUMENT_SUB_ACCOUNT_GLOBAL_NON_CG_NON_BLANK_SECTION,
+                    CUKFSConstants.SUB_ACCOUNT_GLOBAL_CG_ICR_ACCOUNTS_SECTION_NAME);
+            success = false;
+        }
+        return success;
+    }
+    
+    protected void putFieldErrorWithDisplayNameAddedToMessage(String propertyName, String errorConstant) {
+        putFieldError(propertyName, errorConstant, getDisplayNameForGlobalObjectProperty(propertyName));
+    }
+    
     /**
      * Retrieves the label name for a specific property
      * 
      * @param propertyName - property to retrieve label for (from the DD)
      * @return the label
      */
     protected String getDisplayName(String propertyName) {
```

Comment:
should this function be renamed to getDisplayNameForSubAccountProperty

Reply:
Good point, I'll rename this method accordingly.